ts/Data/Converters/JSONConverter.ts

```diff
@@ -0,0 +1,260 @@
+/* *
+ *
+ *  (c) 2009-2023 Highsoft AS
+ *
+ *  License: www.highcharts.com/license
+ *
+ *  !!!!!!! SOURCE GETS TRANSPILED BY TYPESCRIPT. EDIT TS FILE ONLY. !!!!!!!
+ *
+ *  Authors:
+ *  - Pawel Lysy
+ *
+ * */
+
+'use strict';
+
+/* *
+ *
+ *  Imports
+ *
+ * */
+
+import type DataEvent from '../DataEvent';
+import type DataConnector from '../Connectors/DataConnector';
+
+import DataConverter from './DataConverter.js';
+import DataTable from '../DataTable.js';
+import U from '../../Core/Utilities.js';
+const { merge } = U;
+
+/* *
+ *
+ *  Class
+ *
+ * */
+
+/**
+ * Handles parsing and transforming JSON to a table.
+ *
+ * @private
+ */
+class JSONConverter extends DataConverter {
+    /* *
+     *
+     *  Static Properties
+     *
+     * */
+
+    /**
+     * Default options
+     */
+    protected static readonly defaultOptions: JSONConverter.Options = {
+        ...DataConverter.defaultOptions,
+        data: [],
+        orientation: 'columns'
+    };
+
+    /* *
+     *
+     *  Constructor
+     *
+     * */
+
+    /**
+     * Constructs an instance of the JSON parser.
+     *
+     * @param {JSONConverter.UserOptions} [options]
+     * Options for the JSON parser.
+     */
+    public constructor(
+        options?: JSONConverter.UserOptions
+    ) {
+        const mergedOptions = merge(JSONConverter.defaultOptions, options);
+
+        super(mergedOptions);
+
+        this.options = mergedOptions;
+        this.table = new DataTable();
+    }
+
+    /* *
+     *
+     *  Properties
+     *
+     * */
+
+    private columns: Array<DataTable.Column> = [];
+    private headers: Array<string> = [];
+    private dataTypes: Array<Array<string>> = [];
+
+    /**
+     * Options for the DataConverter.
+     */
+    public readonly options: JSONConverter.Options;
+
+    private table: DataTable;
+
+    /* *
+     *
+     *  Functions
+     *
+     * */
+
+    /**
+     * Initiates parsing of JSON structure.
+     *
+     * @param {JSONConverter.UserOptions}[options]
+     * Options for the parser
+     *
+     * @param {DataEvent.Detail} [eventDetail]
+     * Custom information for pending events.
+     *
+     * @emits JSONConverter#parse
+     * @emits JSONConverter#afterParse
+     */
+    public parse(
+        options: JSONConverter.UserOptions,
+        eventDetail?: DataEvent.Detail
+    ): void {
+
+        const converter = this;
+        options = merge(converter.options, options);
+        const {
+            beforeParse,
+            orientation,
+            firstRowAsNames,
+            columnNames
+        } = options;
+        let data = options.data;
+
+        if (data && beforeParse) {
+            data = beforeParse(data);
+        }
+
+        if (!data) {
+            return;
+        }
```

Comment:
The condition should be above `if (data && beforeParse) {` line.

Reply:
Hm I'm wondering if so, because what if the user uses the beforeParse method, and won't return value? Should it throw runtime error? Maybe so, so I'll replace the order :p